R/mock-object.R

```diff
@@ -0,0 +1,81 @@
+
+#' Create a mock object.
+#'
+#' Mock object's primary use is to record calls that are made on the
+#' mocked function. Optionally values can be passed via \code{...} to
+#' make the mock object return them upon subsequent calls. If no value
+#' is passed in \code{...} then \code{NULL} is returned.
+#'
+#' @param ... Values returned upon subsequent calls.
+#' @param cycle Whether to cycle over the return values. If \code{FALSE},
+#'        will fail if called too many times.
+#' @export
+#' @examples
+#' \dontrun{
+#'   m <- mock(1)
+#'   with_mock(summary = m, {
+#'     expect_equal(summary(iris), 1)
+#'     expect_call(m, 1, summary(iris))
+#'   })
+#' }
+#'
+mock <- function (..., cycle = FALSE) {
+  returns <- list(...)
+  call_no <- 0
+  calls   <- list()
+
+  mock_impl <- function (...)
+  {
+    call_no <<- call_no + 1
+    calls[[call_no]] <<- match.call()
+
+    # TODO record the values passed on each call
+
+    if (length(returns)) {
+      if (call_no > length(returns) && !cycle)
+        fail("too many calls to mock object and cycle set to FALSE")
+      return(returns[[(call_no - 1) %% length(returns) + 1]])
+    }
+
+    invisible(NULL)
+  }
+
+  class(mock_impl) <- 'mock'
+  mock_impl
+}
+
+
+
+#' Expectation: does the given call match the expected?
+#'
+#' @inheritParams expect_that
+#' @param mock A \code{\link{mock}} object.
+#' @param no Number of the call made on the \code{mock} object.
+#' @param expected_call Expected call.
+#' @family expectations
+#' @export
+#' @examples
+#' \dontrun{
+#' expect_call(mock, 1, summary(iris))
+#' }
+expect_call <- function (mock, no, expected_call) {
```

Comment:
I'm not sure we need the new expect_call(). How about implementing `[[.mock()` which looks up a call (basically the first part of your expect_call() )? The rest could be handled with the existing expect_equal(). Perhaps `[.mock()` and `as.list.mock()` would be helpful, too.